File: 152. Maximum Product Subarray/optimal_solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # Rules dictate there will be at least one item.
        minimum = nums[0]
        maximum = nums[0]
        best = nums[0]
        
        for ii in range(1, len(nums)):
            curr = [minimum * nums[ii], maximum * nums[ii], nums[ii]]
            minimum = min(curr)
            maximum = max(curr)
            if maximum > best:
                best = maximum
        
        return best

File: 53. Maximum Subarray/optimal_solution.go
func maxSubArray(nums []int) int {
    curr := 0
    max := nums[0]
    if len(nums) == 1 {
        return max
    }
    
    for ii := 0; ii < len(nums); ii++ {
        curr += nums[ii]
        if curr > max {
            max = curr
        } 
        if curr < 0 {
            curr = 0
        }
    }
    return max
}
